### PR TITLE
fix: Save evm context after errors

### PR DESCRIPTION
**Pull request by DanielSchiavini**
- The state was being wiped after an execution error
- This PR saves the evm context even if an error occurs
- Fixes https://github.com/paradigmxyz/pyrevm/issues/17

### CODE REVIEW

File: src/evm.rs [hunk 1]
```diff
@@ -368,14 +368,15 @@ impl EVM {
...
         self.context.env = Box::new(env);
         let evm_context: EvmContext<DB> =
             replace(&mut self.context, EvmContext::new(DB::new_memory()));
         let (result, evm_context) =
-            call_evm(evm_context, self.handler_cfg, self.tracing, is_static)?;
+            call_evm(evm_context, self.handler_cfg, self.tracing, is_static);
         self.context = evm_context;
-        self.result = Some(result.clone());
-        Ok(result)
+        self.result = result.as_ref().map(|r| r.clone()).ok();
```
Comment (rakita) [inline src/evm.rs:379]:
```suggestion
        self.result = result.clone().ok();
```

File: src/evm.rs [hunk 2]
```diff
@@ -368,14 +368,15 @@ impl EVM {
         }
     }
 
+    #[allow(clippy::useless_asref)]
```
Comment (rakita) [inline src/evm.rs:371]:
```suggestion
```

Not good smell to use those

File: src/evm.rs (hunk 1 above)
Comment (DanielSchiavini) [reply src/evm.rs:379]: This does not work
```
error[E0599]: the method `clone` exists for enum `Result<ExecutionResult, PyErr>`, but its trait bounds were not satisfied
   --> src/evm.rs:379:30
    |
379 |         self.result = result.clone().ok();
    |                              ^^^^^ method cannot be called on `Result<ExecutionResult, PyErr>` due to unsatisfied trait bounds
    |
   ::: /home/schiavini/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.19.2/src/err/mod.rs:28:1
    |
28  | pub struct PyErr {
    | ---------------- doesn't satisfy `pyo3::PyErr: std::clone::Clone`
    |
   ::: /home/schiavini/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:502:1
    |
502 | pub enum Result<T, E> {
    | --------------------- doesn't satisfy `_: Clone`
    |
    = note: the following trait bounds were not satisfied:
            `pyo3::PyErr: std::clone::Clone`
            which is required by `std::result::Result<revm::revm_primitives::ExecutionResult, pyo3::PyErr>: std::clone::Clone`
note: the method `clone` exists on the type `revm::revm_primitives::ExecutionResult`
   --> /home/schiavini/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/clone.rs:160:5
    |
160 |     fn clone(&self) -> Self;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^
help: use the `?` operator to extract the `revm::revm_primitives::ExecutionResult` value, propagating a `Result::Err` value to the caller
    |
379 |         self.result = result?.clone().ok();
    |                             +

Process finished with exit code 2
```

File: src/evm.rs (hunk 2 above)
Comment (DanielSchiavini) [reply src/evm.rs:371]: yeah but the suggestion given by clippy does not work (see above)

**Review by charles-cooper — APPROVED** [review]
approved per @rakita offline

**Pull request merged**